File: Dynamic Programming/DP on numbers/Champagne Tower/memoization_app.cpp
#include<bits/stdc++.h>
using namespace std;

//Time complextiy :- O(row*row)
// Space complexity :- O(row*row)

// Approach :-   
// for finding the champagne in the given glass, we can find 
// how much (row-1, col) and (row-1, col+1) is contribution to the given glass 
// which will become a sub-probelm and can be solve using recursion 
// reset the value of sub-cases to 0.0 if negative value is returned 
// the base case will be when an invalid case comes then return 0
// other we reach top of the tower then return poured 

// Link :- https://leetcode.com/problems/champagne-tower/

class Solution {
private:
    double solve(int poured, int row, int col, vector<vector<double>> &dp){
        // invalid cell value 
        if(row < 0 || col < 0 || row < col)
            return 0.0;
        
        // top of the tower 
        if(row == 0 && col == 0)
            return poured;
        
        // memoization check 
        if(dp[row][col] != -1.0)
            return dp[row][col];
        
        // find the left and right value of tower from which current glass will get 
        double left = (solve(poured, row-1, col-1, dp) - 1) / 2.0;
        double right = (solve(poured, row-1, col, dp) - 1) / 2.0;

        // reset value to 0.0 if we get -ve value 
        if(left < 0)        left = 0.0;
        if(right < 0)       right = 0.0;

        // total value will be sum of both left and right tower 
        return dp[row][col] = left + right;
    }

public:
    double champagneTower(int poured, int query_row, int query_glass) {
        vector<vector<double>> dp(query_row+1, vector<double> (query_glass+1, -1.0));
        // use recursion to find the value 
        return min<double>(1.0, solve(poured, query_row, query_glass, dp));
    }
};

File: Dynamic Programming/DP on block/Maximum Path Score in a Grid/memoization_app.java
import java.util.*

// Time complexity :- O(m*n*k)
// Space complexity :- O(m*n*k)

// Approach :- 
// we've to try all the possibilities, to find the maximum value of grid 
// (row, col, cost) will be changing parameter 
// all possibilities, move left and down with valid cost 
// choose the maximum possible value from both case 
// base case will be when we reach at last grid or when cost becomes more than k 
// memoize the solution using 3d dp 

// Links :- https://leetcode.com/problems/maximum-path-score-in-a-grid/

class Solution {
    private int solve(int row, int col, int cost, int[][] grid, int k, int[][][] dp) {
        int val = grid[row][col];
        
        // Calculate cost and score for the CURRENT cell
        int currCost = (val == 1 || val == 2) ? 1 : 0;
        int currScore = (val == 1) ? 1 : (val == 2 ? 2 : 0);

        // If taking this cell puts us over budget, this path is invalid
        if (cost + currCost > k) return -100000;

        // Base case: Reached the start
        if (row == 0 && col == 0) {
            return currScore;
        }

        // Memoization check
        if (dp[row][col][cost] != -1) {
            return dp[row][col][cost];
        }

        int up = -100000;
        int left = -100000;

        // Move Left
        if (col - 1 >= 0) {
            left = solve(row, col - 1, cost + currCost, grid, k, dp);
        }
        // Move Up
        if (row - 1 >= 0) {
            up = solve(row - 1, col, cost + currCost, grid, k, dp);
        }

        // The score of this cell + the best score from the remaining path
        int result = Math.max(left, up) + currScore;
        return dp[row][col][cost] = result;
    }

    public int maxPathScore(int[][] grid, int k) {
        int m = grid.length;
        int n = grid[0].length;
        
        // Correct dimensions: [rows][cols][cost]
        int[][][] dp = new int[m][n][k + 1];
        for (int[][] mat : dp) {
            for (int[] row : mat) {
                Arrays.fill(row, -1);
            }
        }

        int ans = solve(m - 1, n - 1, 0, grid, k, dp);
        
        // If ans is still a large negative number, no path was found within budget k
        return (ans < 0) ? -1 : ans;
    }
}